### PR TITLE
[Dance Diffusion] Add dance diffusion

### PR DESCRIPTION
**Pull request by patrickvonplaten**
cc @apolinario to monitor progress

Checkpoints are uploaded here: https://huggingface.co/harmonai

Maestro Pipeline can be tested with:

```python
from diffusers import DiffusionPipeline
import scipy.io.wavfile

pipe = DiffusionPipeline.from_pretrained("harmonai/maestro-150k")
pipe = pipe.to("cuda")

audios = pipe(num_inference_steps=100, sample_length_in_s=4.0).audios

scipy.io.wavfile.write("maestro_test.wav", pipe.unet.sample_rate, audios)
```

It relies on the `DanceDiffusionPipeline` the `IPNDMScheduler` and the `UNet1DModel` classes.

## TODO
- [x] Convert model weights and successfully port model
- [x] Add scheduler
- [x] Add pipeline

### CODE REVIEW

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 from ..modeling_flax_utils import FlaxModelMixin
 from ..utils import BaseOutput
 from .embeddings_flax import FlaxTimestepEmbedding, FlaxTimesteps
-from .unet_blocks_flax import (
+from .unet_blocks_2d_flax import (
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition_flax.py:13]:
```suggestion
from .unet_blocks_flax import (
```

File: src/diffusers/models/unet_1d_blocks.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+                nn.GroupNorm(1, c_out) if not is_last else nn.Identity(),
+                nn.GELU() if not is_last else nn.Identity(),
+            ]
+        )
+
+    def forward(self, input):
+        return self.main(input) + self.skip(input)
+
+
+def get_down_block(down_block_type, c, c_prev):
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_1d_blocks.py:120]: @natolambert - similar to unet_2d_blocks we'll have a new `unet_1d_blocks.py` file where you can define very customizable unet classes

**Comment by natolambert** [comment]
This model is in concurrent development with #105.

File: src/diffusers/models/unet_1d.py [hunk 3]
```diff
@@ -70,8 +70,9 @@ def __init__(
         self.sample_size = sample_size
 
         # time
-        self.time_proj = GaussianFourierProjection(embedding_size=8)
-        del self.time_proj.W
+        self.time_proj = GaussianFourierProjection(
```
Comment (natolambert) [inline src/diffusers/models/unet_1d.py:73]: Should we format this like in the 2d class?

Also, no embedding after projection?

https://github.com/huggingface/diffusers/blob/6cbb73b57e6b679d0de8ae47e80d50874ad5fb14/src/diffusers/models/unet_2d.py#L101-L109

File: src/diffusers/models/unet_1d.py [hunk 4]
```diff
@@ -132,24 +133,24 @@ def forward(
             otherwise a `tuple`. When returning a tuple, the first element is the sample tensor.
         """
         # 1. time
-        timestep_embed = self.time_proj(timestep[:, None])[..., None].repeat([1, 1, sample.shape[2]])
+        timestep_embed = self.time_proj(timestep)[..., None]
+        timestep_embed = timestep_embed.repeat([1, 1, sample.shape[2]])
 
         sample = torch.cat([sample, timestep_embed], dim=1)
```
Comment (natolambert) [inline src/diffusers/models/unet_1d.py:139]: This logic should maybe go in the block rather than the forward?

File: src/diffusers/models/unet_1d.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            timestep (`torch.FloatTensor` or `float` or `int): (batch) timesteps
+            return_dict (`bool`, *optional*, defaults to `True`):
+                Whether or not to return a [`~models.unet_2d.UNet1DOutput`] instead of a plain tuple.
+
+        Returns:
+            [`~models.unet_2d.UNet1DOutput`] or `tuple`: [`~models.unet_2d.UNet1DOutput`] if `return_dict` is True,
+            otherwise a `tuple`. When returning a tuple, the first element is the sample tensor.
+        """
+        # 1. time
+        timestep_embed = self.time_proj(timestep)[..., None]
```
Comment (natolambert) [inline src/diffusers/models/unet_1d.py:136]: Let's remove indexing.

File: src/diffusers/models/unet_1d.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/models/unet_1d.py]: Think as soon as we have more than GaussianFourier, let's do it - before it's maybe not necessary

File: tests/test_models_unet_1d.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+from diffusers import UNet1DModel
+from diffusers.utils import slow, torch_device
+
+
+torch.backends.cuda.matmul.allow_tf32 = False
+
+
+class UnetModel1DTests(unittest.TestCase):
+    @slow
+    def test_unet_1d_maestro(self):
```
Comment (patrickvonplaten) [inline tests/test_models_unet_1d.py:29]: @natolambert this test needs to pass

File: src/diffusers/models/unet_1d.py (hunk 3 above)
Comment (natolambert) [reply src/diffusers/models/unet_1d.py]: The RL unet1d is different, was prepping for that.

File: src/diffusers/models/unet_1d.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+from ..utils import BaseOutput
+from .embeddings import GaussianFourierProjection, TimestepEmbedding, Timesteps
+from .unet_1d_blocks import get_down_block, get_mid_block, get_out_block, get_up_block
+
+
+@dataclass
+class UNet1DOutput(BaseOutput):
+    """
+    Args:
+        sample (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_1d.py:18]:
```suggestion
        sample (`torch.FloatTensor` of shape `(batch_size, sample_size, num_channels)`):
```

**Comment by patrickvonplaten** [comment]
Maestro Pipeline can be tested with:

```python
from diffusers import DiffusionPipeline
import scipy.io.wavfile

pipe = DiffusionPipeline.from_pretrained("harmonai/maestro-150k")
pipe = pipe.to("cuda")

audios = pipe(num_inference_steps=100, sample_size=2 * pipe.unet.sample_size).audios

scipy.io.wavfile.write("maestro_test.wav", pipe.unet.sample_rate, audios)
```

File: src/diffusers/pipelines/dance_diffusion/pipeline_dance_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+from ...pipeline_utils import AudioPipelineOutput, DiffusionPipeline
+
+
+class DanceDiffusionPipeline(DiffusionPipeline):
+    r"""
+    This model inherits from [`DiffusionPipeline`]. Check the superclass documentation for the generic methods the
+    library implements for all the pipelines (such as downloading or saving, running on a particular device, etc.)
+
+    Parameters:
+        unet ([`UNet2DModel`]): U-Net architecture to denoise the encoded image.
```
Comment (natolambert) [inline src/diffusers/pipelines/dance_diffusion/pipeline_dance_diffusion.py:34]:
```suggestion
        unet ([`UNet1DModel`]): U-Net architecture to denoise the encoded image.
```

File: src/diffusers/schedulers/scheduling_ipndm.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+import torch
+
+from ..configuration_utils import ConfigMixin, register_to_config
+from .scheduling_utils import SchedulerMixin, SchedulerOutput
+
+
+class IPNDMScheduler(SchedulerMixin, ConfigMixin):
+    """
+    Pseudo numerical methods for diffusion models (PNDM) proposes using more advanced ODE integration techniques,
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ipndm.py:26]: Can we add a short comment on why this is different from vanilla pndm?

File: src/diffusers/models/unet_1d_blocks.py [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+    raise ValueError(f"{up_block_type} does not exist.")
+
+
+def get_mid_block(mid_block_type, in_channels, mid_channels, out_channels):
+    if mid_block_type == "UNetMidBlock1D":
+        return UNetMidBlock1D(in_channels=in_channels, mid_channels=mid_channels, out_channels=out_channels)
+    raise ValueError(f"{mid_block_type} does not exist.")
+
+
+class UNetMidBlock1D(nn.Module):
```
Comment (natolambert) [inline src/diffusers/models/unet_1d_blocks.py:191]: How useful is it for me to add a num_layers arg here / clean up blocks?
Could definitely reduce the amount of classes, but not sure if it's worth the time.

File: src/diffusers/models/unet_1d.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+        self.up_blocks = nn.ModuleList([])
+        self.out_block = None
+
+        # down
+        output_channel = in_channels
+        for i, down_block_type in enumerate(down_block_types):
+            input_channel = output_channel
+            output_channel = block_out_channels[i]
+
+            if i == 0:
```
Comment (natolambert) [inline src/diffusers/models/unet_1d.py:102]: Can we make this in the config / saved model?

File: src/diffusers/models/unet_1d.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        sample (`torch.FloatTensor` of shape `(batch_size, sample_size, num_channels)`):
+            Hidden states output. Output of last layer of model.
+    """
+
+    sample: torch.FloatTensor
+
+
+class UNet1DModel(ModelMixin, ConfigMixin):
+    r"""
+    UNet1DModel is a 2D UNet model that takes in a noisy sample and a timestep and returns sample shaped output.
```
Comment (zqevans) [inline src/diffusers/models/unet_1d.py:27]: All the comments here refer to 2D models

File: scripts/convert_dance_diffusion_to_diffusers.py [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+
+    if args.save:
+        pipe.save_pretrained(args.checkpoint_path)
+
+    print("Diff sum", diff_sum)
+    print("Diff max", diff_max)
+
+    assert diff_max < 1e-3, f"Diff max: {diff_max} is too much :-/"
+
+    print(f"Converion for {model_name} succesful!")
```
Comment (zqevans) [inline scripts/convert_dance_diffusion_to_diffusers.py:326]: Typo: Conver**s**ion

Comment (zqevans) [reply scripts/convert_dance_diffusion_to_diffusers.py:326]: also succes**s**ful

File: src/diffusers/models/unet_1d.py [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    ) -> Union[UNet1DOutput, Tuple]:
+        r"""
+        Args:
+            sample (`torch.FloatTensor`): (batch, channel, height, width) noisy inputs tensor
+            timestep (`torch.FloatTensor` or `float` or `int): (batch) timesteps
+            return_dict (`bool`, *optional*, defaults to `True`):
+                Whether or not to return a [`~models.unet_2d.UNet1DOutput`] instead of a plain tuple.
+
+        Returns:
+            [`~models.unet_2d.UNet1DOutput`] or `tuple`: [`~models.unet_2d.UNet1DOutput`] if `return_dict` is True,
```
Comment (zqevans) [inline src/diffusers/models/unet_1d.py:152]: 2d again

File: src/diffusers/models/unet_1d_blocks.py (hunk 10 above)
Comment (zqevans) [reply src/diffusers/models/unet_1d_blocks.py:191]: I wish I'd had a configurable number of layers in my code, so that would be great to have here

Comment (patrickvonplaten) [reply src/diffusers/models/unet_1d_blocks.py:191]: Yeah that's a good point - at the moment I would say maybe wait a bit how the API will change with more 1D models popping and if all of them have **many** layers then let's introduce a more complex config parameter?

File: src/diffusers/pipelines/dance_diffusion/pipeline_dance_diffusion.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        scheduler ([`SchedulerMixin`]):
+            A scheduler to be used in combination with `unet` to denoise the encoded image. Can be one of
+            [`DanceDiffusionScheduler`], or [`DDIMScheduler`].
```
Comment (patil-suraj) [inline src/diffusers/pipelines/dance_diffusion/pipeline_dance_diffusion.py:37]: is there a `DanceDiffusionScheduler` ?

File: src/diffusers/pipelines/dance_diffusion/pipeline_dance_diffusion.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+        if sample_size % down_scale_factor != 0:
+            sample_size = ((sample_length_in_s * self.unet.sample_rate) // down_scale_factor + 1) * down_scale_factor
+            raise logger.info(
+                f"{sample_length_in_s} is increased to {sample_size / self.unet.sample_rate} so that it can be handled"
+                f" by the model. It will be cut to {original_sample_size / self.unet.sample_rate} after the denoising"
+                " process."
+            )
+        sample_size = int(sample_size)
+
+        audio = torch.randn((batch_size, self.unet.in_channels, sample_size), generator=generator, device=self.device)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/dance_diffusion/pipeline_dance_diffusion.py:94]: What's the expected shape here, `channels_last` or `channels_first` ?
In the `UNet1DOutput` the shape is documented as `(bs, sample_size, num_channels)`

File: src/diffusers/schedulers/scheduling_ipndm.py (hunk 9 above)
Comment (patil-suraj) [reply src/diffusers/schedulers/scheduling_ipndm.py:26]: +1

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for adding this model! Looks very good, just left some comments mostly related to docs.

File: tests/pipelines/dance_diffusion/test_dance_diffusion.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        audio_from_tuple_slice = audio_from_tuple[0, -3:, -3:]
+
+        assert audio.shape == (1, 2, self.dummy_unet.sample_size)
+        expected_slice = np.array([-0.7265, 1.0000, -0.8388, 0.1175, 0.9498, -1.0000])
+        assert np.abs(audio_slice.flatten() - expected_slice).max() < 1e-2
+        assert np.abs(audio_from_tuple_slice.flatten() - expected_slice).max() < 1e-2
+
+
+@slow
+@unittest.skipIf(torch_device == "cpu", "Stable diffusion is supposed to run on GPU")
```
Comment (anton-l) [inline tests/pipelines/dance_diffusion/test_dance_diffusion.py:76]:
```suggestion
@require_torch_gpu
```

File: tests/pipelines/dance_diffusion/test_dance_diffusion.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+# limitations under the License.
+
+import gc
+import unittest
+
+import numpy as np
+import torch
+
+from diffusers import DanceDiffusionPipeline, IPNDMScheduler, UNet1DModel
+from diffusers.utils import slow, torch_device
```
Comment (anton-l) [inline tests/pipelines/dance_diffusion/test_dance_diffusion.py:23]:
```suggestion
from diffusers.utils import slow, torch_device
from diffusers.utils.testing_utils import require_torch_gpu
```

**Review by anton-l — APPROVED** [review]

File: src/diffusers/models/unet_1d.py [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+        sample_size (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`, *optional*):
+            Input sample size.
+        in_channels (`int`, *optional*, defaults to 3): Number of channels in the input image.
+        out_channels (`int`, *optional*, defaults to 3): Number of channels in the output.
+        time_embedding_type (`str`, *optional*, defaults to `"positional"`): Type of time embedding to use.
+        freq_shift (`int`, *optional*, defaults to 0): Frequency shift for fourier time embedding.
+        flip_sin_to_cos (`bool`, *optional*, defaults to :
+            obj:`False`): Whether to flip sin to cos for fourier time embedding.
+        down_block_types (`Tuple[str]`, *optional*, defaults to :
+            obj:`("DownBlock2D", "AttnDownBlock2D", "AttnDownBlock2D", "AttnDownBlock2D")`): Tuple of downsample block
```
Comment (pcuenca) [inline src/diffusers/models/unet_1d.py:42]:
```suggestion
            obj:`("DownBlock1D", "DownBlock1DNoSkip", "AttnDownBlock1D")`): Tuple of downsample block
```

File: src/diffusers/models/unet_1d.py [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+            types.
+        up_block_types (`Tuple[str]`, *optional*, defaults to :
+            obj:`("AttnUpBlock2D", "AttnUpBlock2D", "AttnUpBlock2D", "UpBlock2D")`): Tuple of upsample block types.
```
Comment (pcuenca) [inline src/diffusers/models/unet_1d.py:45]:
```suggestion
            obj:`("UpBlock1D", "UpBlock1DNoSkip", "AttnUpBlock1D")`): Tuple of upsample block types.
```

File: src/diffusers/models/unet_1d.py [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+        norm_eps (`float`, *optional*, defaults to `1e-5`): The epsilon for the normalization.
+    """
+
+    @register_to_config
+    def __init__(
+        self,
+        sample_size: int = 65536,
+        sample_rate: Optional[int] = None,
+        in_channels: int = 2,
+        out_channels: int = 2,
```
Comment (pcuenca) [inline src/diffusers/models/unet_1d.py:63]: See @patil-suraj's comments above about number of channels.

**Review by pcuenca — APPROVED** [review]

File: src/diffusers/models/unet_1d.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/diffusers/models/unet_1d.py:102]: yeah definitely! Can we do this in your PR maybe?

**Comment by patrickvonplaten** [comment]
Ran the whole slow tests suite and everything passed
